[immanuelhume] iP

immanuelhume opened the pull request:
# Chungus

> "It's Morbin Time." - Dr. Michael Morbius

*Chungus* (not to be confused with [Big Chungus](http://artsatmichigan.umich.edu/ink/2019/01/11/who-is-big-chungus/)) is a simple task tracking assistant. Chungus is

* Text-based
* Easy to learn
* Written in **Java 11**, one of the languages of all time :weary: 

## Basic Usage

1. Download the latest release [here](https://github.com/immanuelhume/ip/releases)
2. Ensure Java 11 is installed
3. Run *Chungus* with `java -jar chungus.jar`

## Reading the Code

The entry point can be found in `Chungus.java`.

```java
class Chungus {
    ...
    public static void main(String[] args) {
        new Chungus(System.in, System.out, DEFAULT_DB_PATH).spin();
    }
    ...
}
```

Each Chungus instance expects an input and output stream, as well as the path to a file on disk to store its data. By default, it uses `./chungus.db`, relative to the current working directory.

## Feature Roadmap

- [x] Manage different kinds of tasks - todo, deadlines, and events
- [ ] Cure depression (coming soon)

Commit:
Add Gradle support

immanuelhume committed:
Add skeletal code

immanuelhume committed:
Add task saving and retrieval

immanuelhume committed:
Add "mark as done" feature

immanuelhume committed:
Add more kinds of tasks

immanuelhume committed:
Add some simple text testing

Also renamed Duke.java to Chungus.java and created a new Task.java file.

immanuelhume committed:
Add custom exception class

immanuelhume committed:
Add delete task command

immanuelhume committed:
Add save to file functionality

immanuelhume committed:
Replace Strings with LocalDateTime in tasks

Also organized imports

immanuelhume committed:
Merge branch 'branch-Level-7'

immanuelhume committed:
Merge branch 'branch-Level-8'

Also resolved some conflicts and issues with marshalling datetimes

immanuelhume committed:
Use 4 spaces for indentation

immanuelhume committed:
Extract Ui and Parser into their own classes

immanuelhume committed:
Add a formatter config

immanuelhume committed:
Create a TaskList class

Does not provide much benefit over a regular ArrayList for now

immanuelhume committed:
Extract parsing and handling logic to own classes

immanuelhume committed:
Fix indentation

It was overriden to tabs somehow

immanuelhume committed:
Create chungus package

immanuelhume committed:
Merge remote-tracking branch 'origin/add-gradle-support'

immanuelhume committed:
Edit paths in build.gradle

immanuelhume committed:
Clean up exception handling

immanuelhume committed:
Add some unit tests

Also created a new chungus.util package

immanuelhume committed:
Fix wrong handler for unmark command

immanuelhume committed:
Change jar file name

immanuelhume committed:
Add JavaDoc

immanuelhume committed:
Edit code to follow CS2103T convention

immanuelhume committed:
Add command to find task by keyword

immanuelhume committed:
Merge branch 'branch-A-JavaDoc'

immanuelhume committed:
Merge branch 'branch-A-CodingStandard'

immanuelhume committed:
Merge branch 'branch-Level-9'

immanuelhume committed:
Add missing JavaDoc

immanuelhume committed:
Make some small formatting changes

immanuelhume committed:
Add checkstyle plugin

immanuelhume committed:
Restore missing delete command from parser

Also added some more JavaDocs

immanuelhume committed:
Fix build config for VSCode

immanuelhume committed:
Update checkstyle config

immanuelhume committed:
Make Ui abstract

Now each Chungus instance can use any Ui instance. Later, when adding a
GUI in JavaFX, our lives would be easier.

immanuelhume committed:
Add JavaFX interface

immanuelhume committed:
Merge branch 'branch-Level-10'

Lanreath reviewed (commented):
Appreciate the commitment to chonciness 👏 👏

immanuelhume committed:
Add some assertions

We mainly assert some facts about the sizes of strings when
unmarshalling data.

Also, the static methods on the Handler class was renamed to sound more
like a verb.

immanuelhume committed:
Enable assertions for gradlew run

immanuelhume committed:
Rename Ui to NonBlockingUi

This better describes its functionality, since it is important that
the "init" method does not block. Callers are expected to block the
application in their own event loop.

immanuelhume committed:
Replace magic numbers with constants

We also do a check that the dimensions in fxml match those expected
in code.

immanuelhume committed:
Merge pull request #1 from immanuelhume/branch-A-Assertions

Add some assertions

immanuelhume committed:
Merge branch 'master' into branch-A-CodeQuality

immanuelhume committed:
Merge pull request #2 from immanuelhume/branch-A-CodeQuality

Improve code quality

immanuelhume committed:
Add github action

immanuelhume committed:
Update actions config to cache Gradle dependencies

immanuelhume committed:
Use ubuntu only for actions

Because I want to save my free tier minutes.

immanuelhume committed:
Add tagging feature

immanuelhume committed:
Merge branch 'feat-tagging'

ginloy commented on src/main/java/chungus/Task.java:
```diff
@@ -0,0 +1,405 @@
+/**
+ * A todo.
+ */
+class Todo extends Task {
```
Perhaps put the different tasks in separate files? And then put them all in one separate package.

ginloy commented on src/main/java/chungus/util/Pair.java:
```diff
@@ -0,0 +1,51 @@
+ * A general pair class.
+ */
+public class Pair<S, T> {
+    private S first;
+    private T second;
```
Not essential but consider making these private attributes final since there are no setters.

immanuelhume replied:
Good idea

ginloy commented on src/main/java/chungus/Parser.java:
```diff
@@ -0,0 +1,129 @@
+     * @return An appropriate handler for the command.
+     * @throws ChungusException When the expected command format is wrong.
+     */
+    public static Handler parse(String cmd) {
```
Very long method. Perhaps abstract away the parsing so that it is easier to maintain and add new cases. But nice use of switch cases!

immanuelhume replied:
I think this is fine for now as the length comes from the number of cases. Within each case we only extract some values and pass it to the handler so it's not actually doing much. Thanks for the suggestion though!

ginloy commented on src/main/java/chungus/Chungus.java:
```diff
@@ -0,0 +1,98 @@
+ * A task management app.
+ */
+public class Chungus {
+    private NonBlockingUi ui;
+    private TaskList tasks;
+    private Storage db;
+    private AtomicBoolean isRunning;
+
+    public static final String DEFAULT_DB_PATH = System.getProperty("user.dir") + "/chungus.db";
```
Perhaps the static attributes should be above the private final attributes according to coding standards

immanuelhume replied:
In which page did you find this? I wasn't able to find it at https://se-education.org/guides/conventions/java/intermediate.html

ginloy replied:
Tbh idk someone suggested that to me too.

immanuelhume replied:
lol

ginloy commented on src/main/java/chungus/ChungusException.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * An exception to represent whenever a requested task could not be found.
+ */
+class TaskNotFoundException extends ChungusException {
+    public TaskNotFoundException(int idx) {
+        super(String.format("Task %d does not exist", idx));
+    }
+}
+
+/**
+ * An exception to represent when the serialization of a task is invalid.
+ */
+class TaskMarshalException extends ChungusException {
+    public TaskMarshalException(String s) {
+        super(String.format("Bad marshal format: \"%s\"", s));
+    }
+}
```
Same as tasks, maybe put these into separate files and put all the exceptions into one package.

ginloy commented on src/main/java/chungus/Task.java:
```diff
@@ -0,0 +1,405 @@
+ * Represents a task.
+ */
+abstract class Task {
+    private String desc;
+    private boolean isDone;
+
+    protected HashSet<String> tags;
+
+    protected static final DateTimeFormatter DATETIME_FMT = DateTimeFormatter.ISO_LOCAL_DATE_TIME;
```
Same here, perhaps the static attributes should be above the private final attributes according to coding standards

ginloy reviewed (commented):
Overall code looks gr8 with very detailed documentation. Just some minor formatting and organization issues. Consider separating the io, exceptions, tasks into different packages.

immanuelhume committed:
Make pair internals final

immanuelhume committed:
Fix style issues

immanuelhume committed:
Add user guide

immanuelhume committed:
Remove unnecessary eclipse config file

immanuelhume committed:
Update user guide

immanuelhume committed:
Add a window title

immanuelhume committed:
Remove duplicate block in user guide

immanuelhume committed:
Update user guide

Just some rephrasing and disambiguation.

immanuelhume committed:
Add missing bold font
